devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,224 @@
+    'alb_count',
+    'alb_failed',
+    'alb_pushed',
+    'sbe_count',
+    'sbe_balanced',
+    'sbe_pushed',
+    'sbf_count',
+    'sbf_balanced',
+    'sbf_pushed',
+    'ttwu_wake_remote',
+    'ttwu_move_affine',
+    'ttwu_move_balance'
+]
+
+CPU_MEASURES = [
+    'yld_count',
+    'legacy_always_zero',
+    'schedule_count',
+    'sched_goidle',
+    'ttwu_count',
+    'ttwu_local',
+    'rq_cpu_time',
+    'run_delay',
+    'pcount'
+]
+
+class SchedstatsInstrument(Instrument):
+    """
+    An instrument for parsing Linux's schedstats
+
+    Creates a *site* for each CPU and each sched_domain (i.e. for each line of
+    /proc/schedstat), and a *channel* for each item in the schedstats file. For
+    example a *site* named "cpu0" will be created for the scheduler stats on
+    CPU0 and a *site* named "cpu0domain0" will be created for the scheduler
+    stats on CPU0's first-level scheduling domain.
+
+    For example:
+
+     - If :method:`reset` is called with ``sites=['cpu0']`` then all
+       stats will be collected for CPU0's runqueue, with a channel for each
+       statistic.
+
+     - If :method:`reset` is called with ``kinds=['alb_pushed']`` then the count
+       of migrations successfully triggered by active_load_balance will be
+       colelcted for each sched domain, with a channel for each domain.
+
+    The measurements are named according to corresponding identifiers in the
+    kernel scheduler code. The names for ``sched_domain.lb_*`` stats, which are
+    recorded per ``cpu_idle_type`` are suffixed with a ':' followed by the idle
+    type, for example ``'lb_balanced:CPU_NEWLY_IDLE'``.
+
+    Only supports schedstats version 15.
+
+    Only supports the CPU and domain data in /proc/schedstat, not the per-task
+    data under /proc/<pid>/schedstat.
+    """
+
+    mode = INSTANTANEOUS
+
+    sysctl_path = '/proc/sys/kernel/sched_schedstats'
+    schedstat_path = '/proc/schedstat'
+
+    def __init__(self, *args, **kwargs):
+        super(SchedstatsInstrument, self).__init__(*args, **kwargs)
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        # Check schedstats is present in the kernel and the format version
+        # matches what we can parse.
+        try:
+            lines = self.target.read_value(self.schedstat_path).splitlines()
+        except TargetError:
+            if not self.target.file_exists(self.schedstat_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+            raise
+
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # On 4.6+ kernels, schedstats needs to be enabled via kernel cmdline or
+        # sysctl. If not, we'll just get a load of zeroes.
+        self.old_sysctl_value = None
```

Comment:
Just for code readability, maybe by using some "returns" on success/failure, can we move the following checks (lines 128-145) into a dedicated function, something like `_check_availability()` ?